src/java.base/share/classes/jdk/internal/util/Architecture.java

```diff
@@ -0,0 +1,110 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.util;
+
+import jdk.internal.vm.annotation.ForceInline;
+import jdk.internal.vm.annotation.Stable;
+
+/**
+ * System architecture enum values.
+ * Each architecture has a matching {@code public static boolean isXXX()} method
+ * that is true when running on that architecture.
+ * The enum values correspond to architectures as named in the build system.
+ * See the values of `OPENJDK_TARGET_ARCH_OSARCH`.
+ * The order and ordinal values must match the values defined in OperatingSystemProps.java.template.
+ */
+public enum Architecture {
+    X64(),        // Represents AMD64 and X86_64
+    X86(),
+    AARCH64(),
+    RISCV64(),
+    S390(),
+    PPC64LE(),
+    ;
+
+    // Cache a copy of the array for lightweight indexing
+    private static final @Stable Architecture[] archValues = Architecture.values();
+
+    /**
+     * {@return {@code true} if the current architecture is X64, Aka amd64}
+     */
+    @ForceInline
+    public static boolean isX64() {
+        return OperatingSystemProps.TARGET_ARCH_IS_X64;
+    }
+
+    /**
+     * {@return {@code true} if the current architecture is X86}
+     */
+    @ForceInline
+    public static boolean isX86() {
+        return OperatingSystemProps.TARGET_ARCH_IS_X86;
+    }
+
+    /**
+     * {@return {@code true} if the current architecture is RISCV64}
+     */
+    @ForceInline
+    public static boolean isRISCV64() {
+        return OperatingSystemProps.TARGET_ARCH_IS_RISCV64;
+    }
+
+    /**
+     * {@return {@code true} if the current architecture is S390}
+     */
+    @ForceInline
+    public static boolean isS390() {
+        return OperatingSystemProps.TARGET_ARCH_IS_S390;
+    }
+
+    /**
+     * {@return {@code true} if the current architecture is PPC64LE}
+     */
+    @ForceInline
+    public static boolean isPPC64LE() {
+        return OperatingSystemProps.TARGET_ARCH_IS_PPC64LE;
+    }
+
+    /**
+     * {@return {@code true} if the current architecture is AARCH64}
+     */
+    @ForceInline
+    public static boolean isAARCH64() {
+        return OperatingSystemProps.TARGET_ARCH_IS_AARCH64;
+    }
+
+    /**
+     * {@return the current architecture}
+     */
+    public static Architecture current() {
+        return archValues[OperatingSystemProps.CURRENT_ARCH_ORDINAL];
```

Comment:
I think the `Architecture ` is different from the `OperatingSystem`. There are ports of [other architectures](https://github.com/openjdk/jdk/blob/1d517afbd4547171ad6fb6a3356351c2554c8279/make/autoconf/platform.m4#L33-L188) (Including `MIPS64el`, `LoongArch64`, `SPARC v9`, etc) in the upstream.

Will hard coding architectures destroy these ports? Should a "OTHER" or "UNKNOWN" entry be added  to the enum to represent other architectures that are not of concern?

Reply:
There is no benefit to preemptively defining a full set of architectures; we only need those that are used in the OpenJDK runtime selection of options or parameters.
The other and unknown cases can be handled in code using switch as `default` or for `if (xx) {} else {...}` in the else clause.
Ports not supported the OpenJDK can add enum values and the corresponding static `isXXX()` methods if needed.

The template file used in the implementation could be renamed to be more agnostic to OS or arch.

Reply:
I understand that there is no need to define enum entries for all architectures now, but the `current` method seems to cause crashes on other platforms. 

Even worse, `OperatingSystemProps` seems to be unable to compile on other architectures at all:

https://github.com/openjdk/jdk/blob/52ca4a70fc3de9e285964f9545ea8cd54e2d9924/src/java.base/share/classes/jdk/internal/util/OperatingSystemProps.java.template#L40

Does `Architecture` really need to be implemented as an enum? The value of this enum has never been used in this PR except for testing. I think perhaps just providing the isXXX methods is enough.

Reply:
The point of Architecture is reflect the architecture as supported by the build and the runtime mutually and to reflect dependencies on the target hardware. 

What did you use as the example that would not compile on the other architecture?
Did you make the other changes to the build that would be needed for a new architecture?

Reply:
> What did you use as the example that would not compile on the other architecture? 

https://github.com/openjdk/jdk/blob/52ca4a70fc3de9e285964f9545ea8cd54e2d9924/src/java.base/share/classes/jdk/internal/util/OperatingSystemProps.java.template#L40

I don't understand how the above code can be compiled on architectures such as LongArch64, MIPS64el, or ARM32 without modifying the source file.

> Did you make the other changes to the build that would be needed for a new architecture?

I'm not talking about the new architecture, I'm talking about the architecture that OpenJDK already supports. 

According to my understanding, the above code breaks the support of all architectures not listed in `Architecture`. But has `Architecture` really listed all the supported architectures in the OpenJDK mainline?

For example, I want to compile OpenJDK for 32-bit ARM.  I think in this PR, I cannot compile without modifying the source code because `TARGET_ARCH_arm` does not exist. But before this PR, I was able to compile.

Reply:
@Glavo Arm32 builds, though, see GHAs (crossbuilding test). As for the others, as long as the ports have not been integrated into mainline, I expect port maintainers will adapt the enum downstream. They do this anyway.

Reply:
Refactored the way that build time architecture values are mapped to the Architecture enum.
It now uses strings and does not require predefined names in the template file.
Explicitly maps some build time arch values to the corresponding (but not identical Arch enum); for example x86_64 to X64.
Added a method to expose the endian-ness of the build; this allows it to be separated from the architecture.
So for example "ppc64le" maps to PPC64 and the endian-ness is little.
Added an "OTHER" to the enum values so that unrecognized architecture values from the build can be mapped to that enum.
Renamed the template file so it can more naturally cover the os, architecture, endianness, and 64bit'ness of the implementation.
Reworked OperatingSystem to use the same string based interface between the build and the enum.

Reply:
> Refactored the way that build time architecture values are mapped to the Architecture enum. 

Thank you for this, Roger.